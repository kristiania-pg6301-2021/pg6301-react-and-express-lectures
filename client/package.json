{
  "name": "client",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "start": "parcel src/index.html",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-router": "^5.2.0",
    "react-router-dom": "^5.2.0"
  },
  "devDependencies": {
    "parcel": "^2.0.0-beta.2",
    "prettier": "^2.2.1"
  }
}
