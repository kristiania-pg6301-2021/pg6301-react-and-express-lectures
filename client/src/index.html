<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>KristianiaQuiz</title>
  </head>
  <body>
    <div id="app"></div>
  </body>
  <script src="index.jsx"></script>
</html>
