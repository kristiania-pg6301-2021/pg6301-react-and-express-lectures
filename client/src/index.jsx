import React from "react";
import ReactDOM from "react-dom";
import { BrowserRouter, Link } from "react-router-dom";
import { Route, Switch } from "react-router";

function Application() {
  return (
    <BrowserRouter>
      <header>
        <Link to={"/"}>KristianiaQuiz</Link>
      </header>
      <main>
        <Switch>
          <Route path="/login">
            <h1>Login</h1>
          </Route>
          <Route path="/" exact>
            <h1>Home page</h1>
            <ul>
              <li>
                <Link to={"/login"}>Login</Link>
              </li>
            </ul>
          </Route>
        </Switch>
      </main>
    </BrowserRouter>
  );
}

ReactDOM.render(<Application />, document.getElementById("app"));
