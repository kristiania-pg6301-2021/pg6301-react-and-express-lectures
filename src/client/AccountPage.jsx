import React, { useEffect, useState } from "react";

function useLoader(loader, deps) {
  const [loading, setLoading] = useState();
  const [data, setData] = useState();
  const [error, setError] = useState();

  async function reload() {
    setLoading(true);
    setError(undefined);
    setData(undefined);
    try {
      setData(await loader());
    } catch (e) {
      setError(e);
    } finally {
      setLoading(false);
    }
  }

  useEffect(reload, deps || []);

  return { loading, data, error, reload };
}

export function AccountPage() {
  const { loading, data, error } = useLoader(async () => {
    const res = await fetch("/api/account");
    if (!res.ok) {
      throw new Error("Failed to GET " + res.url + ": " + res.statusText);
    }
    return res.json();
  });

  if (loading) {
    return <div>Loading...</div>;
  }

  if (error) {
    return <div>Error: {error.toString()}</div>;
  }

  return (
    <>
      <h1>Account</h1>
      <div>You have ${data.account.balance}</div>
    </>
  );
}
