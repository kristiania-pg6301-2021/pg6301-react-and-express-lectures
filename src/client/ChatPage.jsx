import React, { useEffect, useState } from "react";

function useWsChat() {
  const [chatLog, setChatLog] = useState([]);
  const [ws, setWs] = useState();

  useEffect(() => {
    const ws = new WebSocket("ws://" + window.location.host);
    setWs(ws);
    ws.onopen = (event) => {
      console.log("Opened", event);
    };
    ws.onerror = (event) => {
      console.log(event);
    };
    ws.onmessage = (msg) => {
      console.log(msg);
      const { username, message, id } = JSON.parse(msg.data);
      setChatLog((chatLog) => [...chatLog, { username, message, id }]);
    };
  }, []);

  function sendMessage(json) {
    ws.send(JSON.stringify(json));
  }
  return { chatLog, sendMessage };
}

export function ChatPage({ username }) {
  const { chatLog, sendMessage } = useWsChat();

  function handleSendMessage(message) {
    sendMessage({ username, message });
  }

  return (
    <ChatView
      username={username}
      chatLog={chatLog}
      onSendMessage={handleSendMessage}
    />
  );
}

export function ChatView({ username, chatLog, onSendMessage }) {
  const [message, setMessage] = useState("");

  function handleSubmit(e) {
    e.preventDefault();
    onSendMessage(message);
    setMessage("");
  }

  return (
    <>
      <header>
        <h1>Chat application</h1>
      </header>
      <main>
        <h2>Chat started...</h2>
        <div>Welcome {username}</div>
        <div className={"chatLog"}>
          {chatLog.map(({ id, username, message }) => (
            <div key={id} className={"message"}>
              <strong>{username}:</strong> {message}
            </div>
          ))}
        </div>
      </main>
      <footer>
        <form onSubmit={handleSubmit}>
          <input
            autoFocus={true}
            value={message}
            onChange={(e) => setMessage(e.target.value)}
          />
          <button>Send</button>
        </form>
      </footer>
    </>
  );
}
