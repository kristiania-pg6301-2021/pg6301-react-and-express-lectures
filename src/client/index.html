<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>Hello Kristiania</title>
  </head>
  <body>
    <div id="app"></div>
  </body>
  <script src="index.jsx"></script>
</html>
