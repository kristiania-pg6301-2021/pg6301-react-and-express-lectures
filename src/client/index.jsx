import * as React from "react";
import * as ReactDOM from "react-dom";
import { Application } from "./Application";

ReactDOM.render(<Application />, document.getElementById("app"));
