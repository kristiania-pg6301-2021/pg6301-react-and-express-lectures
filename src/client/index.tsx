import * as React from "react";
import * as ReactDOM from "react-dom";

import {BrowserRouter} from "react-router-dom";
import {Route, Switch} from "react-router";

interface Quiz {
    question: string;
    alternatives: string[];
}


function HomePage() {
    return <>
        <h1>Welcome to the quiz</h1>
        <button>Start new quiz</button>
    </>
}

function QuizPage({quizFactory}: {quizFactory(): Quiz}) {
    const quiz = quizFactory();
    const {question, alternatives} = quiz;
    return <>
        <h1>Quiz</h1>
        <div>Question: {question}</div>
        {alternatives.map(a =>
            <button>{a}</button>
        )}
    </>
}

function App() {
    const quiz: Quiz = {
        question: "Are you happy?",
        alternatives: ["yes", "no", "maybe"]
    }
    return <BrowserRouter>
        <Switch>
            <Route path="/quiz">
                <QuizPage quizFactory={() => quiz} />
            </Route>
            <Route exact path="/">
                <HomePage />
            </Route>
            <Route>
                Not found
            </Route>
        </Switch>
    </BrowserRouter>;
}


ReactDOM.render(<App />, document.getElementById("app"));