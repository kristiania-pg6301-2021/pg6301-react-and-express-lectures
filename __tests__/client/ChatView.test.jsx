import { ChatView } from "../../src/client/ChatPage";
import * as ReactDOM from "react-dom";
import * as React from "react";
import { Simulate } from "react-dom/test-utils";

describe("chat view", () => {
  it("can show existing chat messages", async () => {
    const chatLog = [
      {
        id: "1",
        username: "User 1",
        message: "Hello",
      },
      {
        id: "2",
        username: "User 2",
        message: "Welcome, User 1",
      },
    ];

    const container = document.createElement("div");
    ReactDOM.render(
      <ChatView
        chatLog={chatLog}
        username={"Myself"}
        onSendMessage={jest.fn()}
      />,
      container
    );

    expect(container.querySelector(".chatLog .message").textContent).toEqual(
      "User 1: Hello"
    );
    expect(container).toMatchSnapshot();
  });

  it("can send a new chat message", async () => {
    const container = document.createElement("div");
    const onSendMessage = jest.fn();
    ReactDOM.render(
      <ChatView
        chatLog={[]}
        username={"Myself"}
        onSendMessage={onSendMessage}
      />,
      container
    );
    Simulate.change(container.querySelector("form input"), {
      target: { value: "Hello World" },
    });
    expect(container.querySelector("form input").getAttribute("value")).toEqual(
      "Hello World"
    );
    Simulate.submit(container.querySelector("form"));
    expect(onSendMessage).toBeCalledWith("Hello World");
    expect(container.querySelector("form input").getAttribute("value")).toEqual(
      ""
    );
  });
});
