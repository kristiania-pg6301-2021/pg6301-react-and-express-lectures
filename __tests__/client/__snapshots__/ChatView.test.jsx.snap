// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat view can show existing chat messages 1`] = `
<div>
  <header>
    <h1>
      Chat application
    </h1>
  </header>
  <main>
    <h2>
      Chat started...
    </h2>
    <div>
      Welcome 
      Myself
    </div>
    <div
      class="chatLog"
    >
      <div
        class="message"
      >
        <strong>
          User 1
          :
        </strong>
         
        Hello
      </div>
      <div
        class="message"
      >
        <strong>
          User 2
          :
        </strong>
         
        Welcome, User 1
      </div>
    </div>
  </main>
  <footer>
    <form>
      <input
        value=""
      />
      <button>
        Send
      </button>
    </form>
  </footer>
</div>
`;
